src/lib/webhookserver.js

```diff
@@ -0,0 +1,147 @@
+import {spawn} from 'child_process';
+import Debug from 'debug';
+import Promise from 'promise';
+import http from 'http'
+import path from 'path';
+import waitForPort from '../wait_for_port';
+
+let debug = Debug('docker-worker-runtime:webhookserver');
+
+const DEFAULT_TARGET_PORT = 61022;
+
+class WebhookServer {
+  /**
+   * Create a new instance of webhookserver.
+   * The whclient binary must be in bin-utils.
+   * Default target port is 61022.
+   *
+   * @param {String} binPath - Location of whclient binary
+   * @param {String} clientID - Taskcluster client ID
+   * @param {String} accessToken - Taskcluster access token
+   * @param {String} targetPort - Port to which connections should
+   * forwared
+   */
+  constructor({clientId, accessToken}, targetPort){
+    this.clientId = clientId;
+    this.accessToken = accessToken;
+    this.targetPort = targetPort;
+    if(!clientId || !accessToken){
+      throw "clientId and accessToken are required for WebhookServer";
+    }
+    if(targetPort == null)
+      this.targetPort = DEFAULT_TARGET_PORT;
+
+    this.hooks = {}
+
+    // create a server for handling hooks
+    this.server = http.createServer(this._handler)
+    this.baseURL = '';
+    this.path = path.join(__dirname, "../../bin-utils/whclient")
+  }
+
+  /**
+   * Starts a local server to host hooks and connects to webhooktunnel.
+   * Requests on tunnel endpoint are forwarded to local server.
+   *
+   * @example
+   *
+   * @return {String} - Tunnel endpoint on which worker is hosted
+   */
+  async start() {
+    // start the whclient binary
+    let startClient = new Promise((resolve, reject) => {
+      this.proc = spawn(this.path, 
+        [this.clientId, this.accessToken, this.targetPort],
+        {
+          stdio: ['ignore', 'pipe', 'pipe'],
+        });
+
+      proc.stdout.on('data', (data) => {
+        try{
+          let resp = JSON.parse(data);
+          // if no url in server response
+          if (resp.url === undefined) {
+            return reject('no url in response');
+          }
+          return resolve(resp.url);
+        }catch(e){
+          return reject(e);
+        }
+      });
+
+      proc.stderr.on('data', (data) => {
```

Comment:
If you also use stdout for logging from `whclient` which makes sense... maybe write the url to a file instead...
you can pass a filename by command line argument, and you can use file system watchers to monitor for creation of the file...

Note: when creating the file in whclient, you would have to create it and then atomically rename it... to avoid docker-worker creating an empty file as soon as it's created...